larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -151,181 +249,681 @@ void EventClusterValidationAlgorithm::GetHitParents(
         }
     }
 
+    if (m_handleDeltaRays)
+    {
+        for (auto &[pCaloHit, parents] : hitParents)
+        {
+            parents.m_pMainMC = this->FoldPotentialDeltaRayTo(pCaloHit, parents.m_pMainMC);
+        }
+    }
+
     for (const Cluster *const pCluster : clusters)
     {
         const CaloHitList &isolatedHits{pCluster->GetIsolatedCaloHitList()};
         CaloHitList clusterCaloHits;
         pCluster->GetOrderedCaloHitList().FillCaloHitList(clusterCaloHits);
         clusterCaloHits.insert(clusterCaloHits.end(), isolatedHits.begin(), isolatedHits.end());
+
         for (const CaloHit *const pCaloHit : clusterCaloHits)
         {
-            // Ignoring the calo hit if truth matching is missing
             if (hitParents.find(pCaloHit) == hitParents.end())
+            {
                 continue;
+            }
             hitParents[pCaloHit].m_pCluster = pCluster;
         }
     }
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+const MCParticle* EventClusterValidationAlgorithm::FoldMCTo(const MCParticle *const pMC) const
+{
+    if (!this->IsEM(pMC))
+    {
+        return pMC;
+    }
+
+    bool hasAncestorElectron{false};
+    const MCParticle *pCurrentMC{pMC};
+    const MCParticle *pLeadingMC{pMC};
+    while (!pCurrentMC->IsRootParticle())
+    {
+        const MCParticle *const pParentMC{*(pCurrentMC->GetParentList().begin())};
+        const int parentPdg{std::abs(pParentMC->GetParticleId())};
+        if (parentPdg == PHOTON || parentPdg == E_MINUS)
+        {
+            if (parentPdg == E_MINUS)
+            {
+                hasAncestorElectron = true;
+            }
+            pCurrentMC = pParentMC;
+            continue;
+        }
+        pLeadingMC = pCurrentMC;
+        break;
+    }
+
+    // Don't fold "showers" that consist of only compton scatters
```

Comment:
LArMCParticles do store process information, so in principle you could explicitly look for "only compton scatters" directly. Would this fit your aim here?

Reply:
Good to know, I didnt realise we have access to process information. So I think I could replace this ancestor + child electron counting with explicitly checking the end process type, but I validated this behaves sensibly with event displays so dont think its worth.